### PR TITLE
Eliminate bind2nd

### PR DESCRIPTION
**Pull request by class4kayaker**
Work on issue #2882.

Removes bind2nd, replacing with std::bind

### CODE REVIEW

**Comment by drwells** [comment]
Thank you for doing this extremely dull piece of work.

What you have is fine for C++11, but it won't work in C++03 mode; it looks like you used `std::bind` in one place when you meant to write `std_cxx11::bind` (similarly, use `std_cxx11::placeholders` instead of `std::placeholders`).

This should be good to go once you fix those.

File: include/deal.II/numerics/time_dependent.h [hunk 1]
```diff
@@ -524,7 +525,7 @@ class TimeDependent
    * the TimeStepBase class.
    *
    * Instead of using the above form, you can equally well use
-   * <tt>bind2nd(mem_fun1(&X::unary_function), arg)</tt> which lets the @p
+   * <tt>std_cxx11::bind(&X::unary_function, objX, arg)</tt> which lets the @p
```
Comment (drwells) [inline include/deal.II/numerics/time_dependent.h]: I am that familiar with bind, but since `&X::unary_function` refers to a member function, don't we need to use `mem_fn` (that is, the C++11 replacement for `mem_fun`, note the `u`) here?

File: source/grid/tria.cc [hunk 2]
```diff
@@ -17,7 +17,6 @@
 #include <deal.II/base/memory_consumption.h>
 #include <deal.II/base/table.h>
 #include <deal.II/base/geometry_info.h>
-#include <deal.II/base/std_cxx11/bind.h>
```
Comment (drwells) [inline source/grid/tria.cc]: why is this removed?

File: source/hp/dof_handler.cc [hunk 3]
```diff
@@ -16,7 +16,6 @@
 #include <deal.II/base/memory_consumption.h>
 #include <deal.II/base/geometry_info.h>
 #include <deal.II/base/thread_management.h>
-#include <deal.II/base/std_cxx11/bind.h>
```
Comment (drwells) [inline source/hp/dof_handler.cc]: same

File: include/deal.II/numerics/time_dependent.h (hunk 1 above)
Comment (class4kayaker) [inline include/deal.II/numerics/time_dependent.h]: I believe I misunderstood the intended use. This has been revised to a working approach.

File: source/grid/tria.cc (hunk 2 above)
Comment (class4kayaker) [inline source/grid/tria.cc]: I had been using std_cxx11::bind, but then second-guessed and went to std::bind. I must have then assumed that I had added this, it has been fixed.

File: include/deal.II/grid/filtered_iterator.h [hunk 4]
```diff
@@ -376,11 +376,11 @@ namespace IteratorFilters
  * @endcode
  * then
  * @code
- *   std::bind2nd (std::ptr_fun(&level_equal_to<active_cell_iterator>), 3)
+ *   std_cxx11::bind (std::ptr_fun(&level_equal_to<active_cell_iterator>), std_cxx11::_1, 3)
```
Comment (tjhei) [inline include/deal.II/grid/filtered_iterator.h]: I don't think ``::bind`` needs ``::ptr_fun`` to work.

File: include/deal.II/numerics/time_dependent.h [hunk 5]
```diff
@@ -283,9 +283,10 @@ template <int dim, int spacedim> class Triangulation;
  *   ...
  *
  *   do_loop (mem_fun (&TimeStepBase_Tria<dim>::init_for_refinement),
- *            bind2nd (mem_fun1 (&TimeStepBase_Wave<dim>::refine_grid),
- *                     TimeStepBase_Tria<dim>::RefinementData (top_threshold,
- *                                                             bottom_threshold)),
+ *            std_cxx11::bind (&TimeStepBase_Wave<dim>::refine_grid,
+ *                             std_cxx11::_1
```
Comment (drwells) [inline include/deal.II/numerics/time_dependent.h]: needs a comma

**Comment by drwells** [comment]
I did some experiments to verify what goes where:

```cpp
#include <deal.II/base/point.h>
#include <deal.II/base/std_cxx11/bind.h>

#include <iostream>

using namespace dealii;

template<int dim>
double foo(const Point<dim> p)
{
  return 42.0 + p[0];
}

int main()
{
  auto bound_norm = std_cxx11::bind(&Point<2>::norm, std_cxx11::_1);
  auto foo_1 = std_cxx11::bind(foo<2>, std_cxx11::_1);
  auto foo_2 = std_cxx11::bind(&foo<2>, std_cxx11::_1);


  Point<2> point {1.0, 1.0};
  std::cout << bound_norm(point) << std::endl;
  std::cout << foo_1(point) << std::endl;
  std::cout << foo_2(point) << std::endl;
}
```

which compiles without warnings and yields the correct output when run.

This works. We do need the `&` in front of the member function `Point<2>::norm`; I suppose `bind` is smart enough to figure out that it is a pointer to a member function. The rules for pointers to plain functions are different (if I recall correctly from C just passing the name of a function implies a pointer to the function) so it looks like the `&` is optional in that case.

As for the line replacing

```cpp
bind2nd(mem_fun1(&X::unary_function), arg)
```

to the best of my knowledge this line doesn't make any sense anyway (`mem_fun1` does not appear in the language standard so I am not sure what the author's intent was) and this function is only ever called with something like

```cpp
init_function (static_cast<typename InitFunctionObject::argument_type>
               (&*timesteps[step]));
```

which implies that we should really use

```cpp
std_cxx11::mem_fn(&X::function_without_arguments)
```

where `mem_fn` is new in C++11 and we need to wrap the boost equivalent.

@class4kayaker You may have gotten more than you bargained for with this patch :) I think the other usages are easier.

**Comment by drwells** [comment]
Another possibility would be to use

```cpp
std_cxx11::bind(&X::function_without_arguments, _1)
```

so that we don't need to wrap `mem_fn`.

I am still guessing how this class works and am only *pretty sure* that I am right. @bangerth does this look reasonable to you? I believe you wrote this class originally.

**Comment by class4kayaker** [comment]
The planned scope for this pull-request was to deal with all of the bind2nd calls, and leave the other deprecated functions for later examination. My understanding of the reason for the std_cxx11::bind use being described there is to allow for variations on the arguments being supplied.

I am currently reading through the class, and I am not sure how it obtains the reference to the object that needs to be used for bound member functions with either bind or mem_fn. There may well be an error, but I cannot see how the currently implemented version differs from what had been there.

**Comment by drwells** [comment]
@class4kayaker that is a fair point and I agree; we should not worry about problems in examples at this time.

I believe that they are copied by value from [this example](http://en.cppreference.com/w/cpp/utility/functional/bind) (the author used `std::cref`) but that is outside the scope of this PR.

While it seems to work in C++11 mode, `grid_refinement.cc` needs our bind wrapper to be included to work correctly in C++03 mode.

**Comment by bangerth** [comment]
I looked through it all and it all looks quite reasonable to me.  @drwells,  if you don't have further objections, please merge!

**Comment by tjhei** [comment]
> I am not sure how it obtains the reference to the object that needs to be used for bound member functions

If the first argument to bind is a member function instead of a global function, the second argument to bind is a reference to the object. The example at http://en.cppreference.com/w/cpp/utility/functional/bind might help to understand it.

**Comment by class4kayaker** [comment]
The question was not how std::bind works, but how the reference is being supplied in this case. I already knew that the second argument in the case of a member function is a reference to the object, but when looking into the do_loop member function that is being referenced in that comment I did not see where the references were being supplied.

**Comment by bangerth** [comment]
`std::bind` takes all of its arguments by value. That means that if you write something like this:
```
  class X { void foo(); }
  X x;
  auto f = std::bind(&X::foo, x);
```
that you actually call the member function *one a copy* of the object `x`, not on `x` itself. If you want to avoid this, you need to say
```
  auto f = std::bind(&X::foo, std::ref(x));
```
Was this the question, or did you have anything else in mind?

**Comment by class4kayaker** [comment]
The question was more that in the do_loop function where the bound function would be used, I was not able to find the references to the object that should be used for the member function call. After taking another look, I see that the reference I was expecting was being supplied. It was mostly an issue due to trying to figure out the intended usage to confirm that the comment being modified was still correct by looking how the argument was being used in the do_loop function. I did just take another look, and I see what I overlooked now. I was confusing a reference to a member variable for another variable, and therefore not noticing that it was actually the reference I was expecting, partially due to looking for/expecting something to be wrong due to the issue being brought up specifically.

**Comment by bangerth** [comment]
OK. So that sounds like you think this is complete and ready to merge?

**Comment by class4kayaker** [comment]
I believe so. Do you want me to squash all of this into a single commit now?

**Comment by bangerth** [comment]
This is fine -- your commits are already nicely separated.
